pkg/scheduler/core/generic_scheduler_test.go

```diff
@@ -1504,3 +1504,90 @@ func TestCacheInvalidationRace(t *testing.T) {
 		t.Errorf("Predicate should have been called twice. Was called %d times.", callCount)
 	}
 }
+
+// TestCacheInvalidationRace2 tests that cache invalidation is correctly handled
+// when an invalidation event happens during predicate is running.
+func TestCacheInvalidationRace2(t *testing.T) {
+	// Create a predicate that returns false the first time and true on subsequent calls.
+	podWillFit := false
+	var (
+		callCount        int
+		cycleStart       = make(chan struct{})
+		cacheInvalidated = make(chan struct{})
+		once             sync.Once
+	)
+	testPredicate := func(pod *v1.Pod,
+		meta algorithm.PredicateMetadata,
+		nodeInfo *schedulercache.NodeInfo) (bool, []algorithm.PredicateFailureReason, error) {
+		callCount++
+		var (
+			fit            bool
+			failureReasons []algorithm.PredicateFailureReason
+		)
+		if !podWillFit {
+			podWillFit = true
```

Comment:
You could shorten this function by moving once.Do to the top and returning inside the conditional statement.

Reply:
Updated.